subproviders/subscriptions.js

```diff
@@ -0,0 +1,131 @@
+const EventEmitter = require('events').EventEmitter
+const FilterSubprovider = require('./filters.js')
+const from = require('../util/rpc-hex-encoding.js')
+const inherits = require('util').inherits
+const utils = require('ethereumjs-util')
+
+function SubscriptionSubprovider(opts) {
+  const self = this
+
+  opts = opts || {}
+
+  EventEmitter.apply(this, Array.prototype.slice.call(arguments))
+  FilterSubprovider.apply(this, [opts])
+
+  this.subscriptions = {}
+}
+
+inherits(SubscriptionSubprovider, FilterSubprovider)
+
+// a cheap crack at multiple inheritance
+// I don't really care if `instanceof EventEmitter` passes...
+Object.assign(SubscriptionSubprovider.prototype, EventEmitter.prototype)
```

Comment:
I think instead of `SubscriptionSubprovider` being an EE, we can just make engine an EE and require that to be passed into the constructor -- and lol directly emit events on it /shrug

Reply:
I'd rather leave it like it is - I see no reason why this class needs to be tightly coupled to `engine` like that. In `ganache-core` I make the `provider` an `EventEmitter` (it must be, per web3 requirements), and I proxy the EE methods for the `data` event directly to the `SubscriptionSubprovider`. I like it this way over proxying to the engine because it's more explicit that I'm subscribing to pub/sub events.

Reply:
Now that I've thought about it a bit more, I wonder if there shouldn't be an event middleware stack. As pub/sub grows in scope, I could imagine wanting to write middleware that would squelch or mutate notifications.

Maybe that's just YAGNI, though...